parcels/kernel.py

```diff
@@ -340,10 +346,17 @@ def execute_python(self, pset, endtime, dt):
                 if res in [StateCode.Success, OperationCode.Delete]:
                     # Update time and repeat
                     p.time += p.dt
+                    if reset_dt and p.dt == pdt_prekernels:
+                        p.dt = dt
                     p.update_next_dt()
                     if analytical:
                         p.dt = np.inf
-                    dt_pos = min(abs(p.dt), abs(endtime - p.time))
+                    if abs(endtime - p.time) < abs(p.dt):
+                        dt_pos = abs(endtime - p.time)
+                        reset_dt = True
+                    else:
+                        dt_pos = abs(p.dt)
+                        reset_dt = False
```

Comment:
Did you try the vectorisation here  that you suggested, @ckehl? Did it not work?

Reply:
I tried the vectorisation suggestion myself - running into numerical accuracy errors and numerical overflows in the `math.ceil(....)` function. So, at the points and tests where it matters, it fails. Probably it could be made working, but the gain doesn't justify then the time investment in this case, and Reint has a working solution - so we better just apply that.

Reply:
OK, let's then keep this code as is

@reint-fischer, can you confirm here in the comments that the unit test breaks on the current master, but passes here (i.e. the PR fixes the bug)?
 
Except for fixing another flake8 issue, I think this PR is then ready for merging?

Reply:
Yes, the test fails on master, and since it is included here and the checks are ticked, it is fixed by the PR.

Reply:
great - seems ready to be merged! (waiting for that eagerly) :D
Did Cleo also green-light the fix ?

Reply:
I still observe some fluctuations in my particle.dt in to me unexpected parts of the simulations, which I cannot necessarily explain. My output dt is already a multiple of my computational dt, and furthermore I have to output particle.dt in very high precision to observe it so smells like a round off somewhere.  However, maybe that is for another discussion and since it doesn't any more grind the simulations to an extreme slow down for me the light is green!